controllers/ratelimitpolicy_controller.go

```diff
@@ -82,39 +85,56 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
 		logger.V(1).Info(string(jsonData))
 	}
 
-	if rlp.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(rlp, rateLimitPolicyFinalizer) {
-		if err := r.finalizeRLP(ctx, rlp); err != nil {
-			return ctrl.Result{}, err
-		}
+	markedForDeletion := rlp.GetDeletionTimestamp() != nil
 
-		logger.Info("removing finalizer")
-		controllerutil.RemoveFinalizer(rlp, rateLimitPolicyFinalizer)
-		if err := r.UpdateResource(ctx, rlp); client.IgnoreNotFound(err) != nil {
+	// fetch the target network object
+	targetObj, err := r.FetchValidTargetRef(ctx, rlp.GetTargetRef(), rlp.Namespace)
+	if err != nil {
+		if !markedForDeletion {
+			if apierrors.IsNotFound(err) {
+				logger.V(1).Info("Network object not found. Cleaning up")
+				delResErr := r.deleteResources(ctx, rlp, nil)
+				if delResErr == nil {
+					delResErr = err
+				}
+				return r.reconcileStatus(ctx, rlp, nil, delResErr)
+			}
 			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, nil
+		targetObj = nil // we need the object set to nil when there's an error, otherwise deleting the resources (when marked for deletion) will panic
 	}
 
-	// Ignore deleted resources, this can happen when foregroundDeletion is enabled
-	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
-	if rlp.GetDeletionTimestamp() != nil {
+	// handle authpolicy marked for deletion
+	if markedForDeletion {
+		if controllerutil.ContainsFinalizer(rlp, rateLimitPolicyFinalizer) {
+			logger.V(1).Info("Handling removal of ratelimitpolicy object")
+
+			if err := r.deleteResources(ctx, rlp, targetObj); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			logger.Info("removing finalizer")
+			if err := r.RemoveFinalizer(ctx, rlp, rateLimitPolicyFinalizer); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
 		return ctrl.Result{}, nil
 	}
 
+	// add finalizer to the ratelimitpolicy
 	if !controllerutil.ContainsFinalizer(rlp, rateLimitPolicyFinalizer) {
 		controllerutil.AddFinalizer(rlp, rateLimitPolicyFinalizer)
 		if err := r.UpdateResource(ctx, rlp); client.IgnoreNotFound(err) != nil {
 			return ctrl.Result{Requeue: true}, err
 		}
 	}
 
-	specResult, specErr := r.reconcileSpec(ctx, rlp)
-	if specErr == nil && specResult.Requeue {
-		logger.V(1).Info("Reconciling spec not finished. Requeueing.")
-		return specResult, nil
-	}
+	// reconcile the ratelimitpolicy spec
+	specErr := r.reconcileResources(ctx, rlp, targetObj)
```

Comment:
missing requeues. The reconcile Resources may want to requeue

Reply:
@eguzki ,

I think there was no case where the `reconcileResources` (previously `reconcileSpec`) returned `Requeue: true`, so the condition we had here before checking it was useless.

Reply:
I do not know if there was a case, but even if there is none, the method allows requeuing, not just returning error. And I think it is useful for some uses cases. Maybe it was in the past, or will be in the future. The `kuadrant-controller` also has this capability.